models/progresses.js

```diff
@@ -0,0 +1,138 @@
+ * @throws {Error} If a progress document has already been created for the given user or task on the current day.
+ **/
+const createProgressDocument = async (progressData) => {
+  const { type, userId, taskId } = progressData;
+  const createdAtTimestamp = new Date().getTime();
+  const currentHourIST = new Date().getUTCHours() + 5.5; // IST offset is UTC+5:30
+  const isBefore6amIST = currentHourIST < 6;
+  const progressDateTimestamp = isBefore6amIST
+    ? new Date().setUTCHours(0, 0, 0, 0) - 24 * 60 * 60 * 1000
+    : new Date().setUTCHours(0, 0, 0, 0);
+
+  if (type === "task") {
+    const { taskData } = await fetchTask(taskId);
+    if (!taskData) {
+      throw new Error(`Task with id ${taskId} does not exist`);
+    }
+  }
+
+  const query =
+    type === "user"
+      ? progressesCollection.where("userId", "==", userId)
+      : progressesCollection.where("taskId", "==", taskId);
+
+  const existingDocumentSnapshot = await query.where("date", "==", progressDateTimestamp).get();
+
+  if (!existingDocumentSnapshot.empty) {
+    throw new Conflict(`${type.charAt(0).toUpperCase() + type.slice(1)} Progress for the day has already been created`);
+  }
+
+  const progressDocument = { ...progressData, createdAt: createdAtTimestamp, date: progressDateTimestamp };
+  const { id } = await progressesCollection.add(progressDocument);
+  return { ...progressDocument, id };
+};
+
+/**
+ * This function retrieves the progress document for a specific user or task, or for all users or all tasks if no specific user or task is provided.
+ * @param queryParams {object} This is the data that will be used for querying. It should be an object that includes key-value pairs for the fields - type, userId, taskId.
+ * @returns {Promise<object>} A Promise that resolves with the progress document objects.
+ * @throws {Error} If the userId or taskId is invalid or does not exist.
+ **/
+const getProgressDocument = async (queryParams) => {
+  const { type, userId, taskId } = queryParams;
+  if (userId) {
+    const { userExists } = await fetchUser({ userId });
+    if (!userExists) {
+      throw new Error(`User with id ${userId} does not exist`);
+    }
+  } else if (taskId) {
+    const { taskData } = await fetchTask(taskId);
+    if (!taskData) {
+      throw new Error(`Task with id ${taskId} does not exist`);
+    }
+  }
+  let query;
+  if (type) {
+    query = progressesCollection.where("type", "==", type);
+  } else {
+    if (userId) {
+      query = progressesCollection.where("type", "==", "user").where("userId", "==", userId);
+    } else if (taskId) {
+      query = progressesCollection.where("type", "==", "task").where("taskId", "==", taskId);
+    }
+  }
+  const progressesDocs = await query.get();
+  const docsData = [];
+  progressesDocs.forEach((doc) => {
+    docsData.push(doc.data());
+  });
+  return docsData;
```

Comment:
This flow does not seem to be scalable to me if we start adding more `queyParams` I think a better way to handle this would be to write the logic related to each query param as its independent function and conditionally call it depending on what is passed to us in request. 

We can also create a common function that depending on the query returns us data (docsData) this can be then used inside the functions we created above for each query param.

Reply:
Done. Please check now. Thanks for the suggestion.